### PR TITLE
[1/n][Adjoint Module] Enable proper JAX support in adjoint solver

### PR DESCRIPTION
**Pull request by smartalecH**
### Stack

Each PR is based on the one above it, so every diff here is just its own delta.

| | PR | |
|---:|---|---|
| 1 | #3277 | Fix MaterialGrid design-variable attribution (#1984) |
| 2 | #3290 | Fix source deposition across chunk boundaries |
| 3 | #3292 | Fix adjoint source weighting on a symmetry plane (#3291) |
| 4 | #3280 | Enable proper JAX support in the adjoint module **&larr; you are here** |
| 5 | #3281 | Differentiable angular-spectrum propagation |
| 6 | #3283 | Adjoint gradients w.r.t. source amplitudes |
| 7 | #3284 | Launch a mode through a stratified stack |
| 8 | #3285 | Adjoint gradients w.r.t. geometry (centre and size) |
| 9 | #3286 | Register a material grid's susceptibilities |
| 10 | #3287 | Evaluate the dispersive adjoint at the discrete lineshape |
| 11 | #3293 | Subpixel-smooth dispersive materials and their shape derivative |
| 12 | #3294 | Scalable Pade extrapolation for DFT monitors (#3217) |
| 13 | #3295 | 3D grating coupler example: fiber, grating and mirror |

This PR makes Jax a "first-class" citizen within Meep's adjoint model. While meep technically supported Jax already, the interface was buggy, lacked several important features, and didn't play nice with the existing autograd engine. This resolves that.

**Allow users to specify their objective function using either autograd's `numpy` or jax's `numpy`**. The backend handles the vjps automatically. You can't mix and match though (e.g. use meep's autograd filters with a jax objective).
```python
from autograd import numpy as npa

def objective(mode_coeff, dft_fields):
    return npa.abs(mode_coeff)**2 + npa.sum(npa.abs(dft_fields)**2, axis=1)
```
OR
```python
from jax import numpy as npj

def objective(mode_coeff, dft_fields):
    return npj.abs(mode_coeff)**2 + npj.sum(npj.abs(dft_fields)**2, axis=1)
```
Meep handles both just fine:
```python
opt = mpa.OptimizationProblem(
    simulation=sim,
    objective_functions=[objective],
    objective_arguments=[mode_monitor, dft_monitor],
    design_regions=[design_region],
    frequencies=frequencies,
)

value, gradient = opt([rho])
```

**Extend `MeepJaxWrapper` to properly support single-frequency objectives and vector-valued (multi-frequency) objectives**. There was a bug in the single-frequency case. The multi-frequency case previously required the user to scalarize the objective (e.g. reduce across the frequency/wavelength dimension). You can now specify a vector-valued objective function (like you can with the standard `OptimizationProblem` API) and jax will return a dense jacobian:

```py
def loss(rho, thickness, tilt):
    (dft,) = wrapped_meep([rho])
    return postprocess(dft, thickness, tilt)   # scalar

value, (d_rho, d_thickness, d_tilt) = jax.value_and_grad(
    loss, argnums=(0, 1, 2))(rho, 1.8, 8.0)
```

**Package optimization parameters as pytrees** (and return the gradient as an organized pytree). The end2end jax workflow allows you to chain operations before the meep simulations (preprocessing) and after (postprocessing). The user may want to collect gradients with parameters that tune these routines too. But bookeeping all of this is a pain. So we adopt the same approach e.g. `fmmax` uses via pytrees. You can pass a single pytree (a Dict/NamedTuple) to your objective function and the gradient routine will _also_ produce a pytree. This also works with vector-valued objectives (e.g. for minimax optimization):

```py
class Stack(NamedTuple):
    thickness: jnp.ndarray     # (3,) oxide, antireflection coating, air
    index: jnp.ndarray         # (3,)

class Params(NamedTuple):
    latent: jnp.ndarray        # (601,) design degrees of freedom
    beta: float                # preprocessing: projection strength
    stack: Stack               # post-processing: the layers above the chip
    tilt: float                # post-processing: fiber tilt, in degrees

def project(x, beta, eta=0.5):
    """The usual tanh projection, written in jax.numpy -- see the note below."""
    return (jnp.tanh(beta * eta) + jnp.tanh(beta * (x - eta))) / (
        jnp.tanh(beta * eta) + jnp.tanh(beta * (1 - eta))
    )

def loss(p):
    rho = project(conic_filter(p.latent), p.beta)   # preprocessing, in jax.numpy
    (ez, hx) = wrapped_meep([rho])
    field = propagate(ez, hx, p.stack)          # a stratified-media propagator, say
    return fiber_overlap(field, p.tilt)         # (nfreq,)

values, jacobian = mpa.value_and_jacobian(loss)(params)
jax.tree_util.tree_map(lambda leaf: leaf.shape, jacobian)
```

**Various other bug fixes and removed antipatterns.**

### CODE REVIEW

**Review by oskooi — APPROVED** [review]
LGTM.

**Comment by stevengj** [comment]
I remember us having  big problems with Jax and MPI on multi-core nodes when @ianwilliamson first implemented jax support.  In particular, JAX tried to grab all of the CPUs on the node, which is bad if you are doing MPI intra-node parallelism.  Is there a way to disable that these days?

**Comment by smartalecH** [comment]
Closing in favour of #3299, which is the same branch and the same content, opened from this repository rather than from a fork.

The reason is purely mechanical: this series is a stack of 13 pull requests, each based on the one below it so that every diff is only its own delta. GitHub's native stacked pull requests cannot include a PR opened from a fork, and four of the thirteen — including this one — were. Rather than leave the series half-stacked, all four have been reopened with their branches hosted here so the whole chain can be linked.

Nothing about the change itself has been revised, and the description is carried over verbatim. The discussion and reviews on this PR stay readable here; please continue on #3299.
